### PR TITLE
Add config_proc_macro

### PR DESCRIPTION
**Pull request by topecongiro**
This PR replaces the current `configuration_option_enum` with `config_type` proc macro.

The long term goal is to make `Config` more maintainable and extensible, and this is the first step.

### CODE REVIEW

File: config_proc_macro/src/attrs.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+pub fn find_doc_hint(attrs: &[syn::Attribute]) -> Option<String> {
```
Comment (scampi) [inline config_proc_macro/src/attrs.rs]: Would be nice to have comments for the public methods!

File: config_proc_macro/src/item_enum.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            }
+        }
+    });
+    quote! {
+        impl ::std::str::FromStr for #ident {
+            type Err = &'static str;
+
+            fn from_str(s: &str) -> Result<Self, Self::Err> {
+                #if_patterns
+                return Err("Bad variant");
```
Comment (scampi) [inline config_proc_macro/src/item_enum.rs]: Maybe add the argument `s` to the returned error in order to have a nicer message.

**Comment by topecongiro** [comment]
Thanks for your review! Yeah I was bit lazy about comments... will do shortly.

**Comment by scampi** [comment]
I understand! I have the same ;o) The procedural macro really improve the code base !

File: config_proc_macro/src/attrs.rs (hunk 1 above)
Comment (topecongiro) [reply config_proc_macro/src/attrs.rs]: Added some comments inhttps://github.com/rust-lang/rustfmt/pull/3484/commits/efa45d2b902945a0c579428eb7d0ecfc8e37e826.

Comment (scampi) [reply config_proc_macro/src/attrs.rs]: @topecongiro thanks a lot for the comprehensive comments!

**Pull request merged**